src/Codecs/V3/V3.jl

```diff
@@ -0,0 +1,531 @@
+    index_location::Symbol      # :start or :end
+end
+name(::ShardingCodec) = "sharding_indexed"
+
+"""
+    JSON.lower(c::ShardingCodec)
+
+Serialize ShardingCodec to JSON format for Zarr v3 metadata.
+"""
+function JSON.lower(c::ShardingCodec)
+    return Dict(
+        "name" => "sharding_indexed",
+        "configuration" => Dict(
+            "chunk_shape" => collect(c.chunk_shape),
+            "codecs" => [JSON.lower(codec) for codec in c.codecs],
+            "index_codecs" => [JSON.lower(codec) for codec in c.index_codecs],
+            "index_location" => string(c.index_location)
+        )
+    )
+end
+
+"""
+    getCodec(::Type{ShardingCodec}, d::Dict)
+
+Deserialize ShardingCodec from JSON configuration dict.
+"""
+function getCodec(::Type{ShardingCodec}, d::Dict)
+    config = d["configuration"]
+    N = length(config["chunk_shape"])
+    chunk_shape = NTuple{N,Int}(config["chunk_shape"])
+    codecs = [getCodec(codec_dict) for codec_dict in config["codecs"]]
+    index_codecs = [getCodec(codec_dict) for codec_dict in config["index_codecs"]]
+    index_location = Symbol(get(config, "index_location", "end"))
+    return ShardingCodec{N}(chunk_shape, codecs, index_codecs, index_location)
+end
+
+const MAX_UINT64 = typemax(UInt64)
+
+"""
+    ChunkShardInfo
+
+Information about a chunk's location within a shard.
+"""
+struct ChunkShardInfo
+    offset::UInt64  # Byte offset within shard where chunk begins
+    nbytes::UInt64  # Number of bytes the chunk occupies
+end
+
+ChunkShardInfo() = ChunkShardInfo(MAX_UINT64, MAX_UINT64)  # Empty chunk marker
+
+"""
+    ShardIndex{N}
+
+Internal structure representing the shard index.
+Stores chunk location info for an N-dimensional grid of chunks.
+Empty chunks are marked with ChunkShardInfo(MAX_UINT64, MAX_UINT64)
+"""
+struct ShardIndex{N}
+    chunks::Array{ChunkShardInfo, N}  # N-dimensional array of chunk info
+end
+
+"""
+    ShardIndex(chunks_per_shard::NTuple{N,Int})
+
+Create an empty shard index with all chunks marked as empty.
+"""
+function ShardIndex(chunks_per_shard::NTuple{N,Int}) where N
+    chunks = fill(ChunkShardInfo(), chunks_per_shard)
+    return ShardIndex{N}(chunks)
+end
+
+"""
+    get_chunk_slice(idx::ShardIndex, chunk_coords::NTuple{N,Int})
+
+Get the byte range (offset, offset+nbytes) for a chunk, or nothing if empty.
+"""
+function get_chunk_slice(idx::ShardIndex, chunk_coords::NTuple{N,Int}) where N
+    info = idx.chunks[chunk_coords...]
+    
+    if info.offset == MAX_UINT64 && info.nbytes == MAX_UINT64
+        return nothing
+    end
+    
+    return (Int(info.offset), Int(info.offset + info.nbytes))
+end
+
+"""
+    set_chunk_slice!(idx::ShardIndex, chunk_coords::NTuple{N,Int}, offset::Int, nbytes::Int)
+
+Set the byte range for a chunk in the index.
+"""
+function set_chunk_slice!(idx::ShardIndex, chunk_coords::NTuple{N,Int}, offset::Int, nbytes::Int) where N
+    idx.chunks[chunk_coords...] = ChunkShardInfo(UInt64(offset), UInt64(nbytes))
+end
```

Comment:
Perhaps we should defer the casting to `UInt64` to the `ChunkShardInfo` constructor.

Also, I think we could relax the arguments to `Integer` rather than `Int`. Also, I am definitely sure `Int` should not depend on the system word size (`Int64` or `Int32`).